### DISCUSSION TITLE
adding variant ships to shipyards

### DISCUSSION THREAD

**Discussion by daggs1** [endless-sky/endless-sky#7124] [General]
### Problem Description

I'm working on some new variant ships feature which I want to add to a specific shipyard.
is there a proper way todo so?

### Related Issue Links

N/A

### Desired Solution

the ship can be purchased from the shipyard

### Alternative Approaches

I can implement the code is it is missing

**Comment by warp-core** [comment]
Yes, you can add ship variants to shipyards. Just use the variant name instead of the base ship name.
Like this:
```
shipyard "whatever"
	"Auxiliary (Cargo)"
	"Auxiliary (Transport)"
	"Auxiliary"
```

The variant name will not currently appear in the shipyard, so, in the example above, all three ships would be labelled `Auxiliary`. #7121 changes this so the variant name will appear.

**Comment by daggs1** [comment]
interesting, I added them under new type of shipyard and I see this:
> Warning: shipyard "Lionheart Custom" is referred to, but has no ships.
and I don't see them in the shipyard

**Comment by warp-core** [comment]
Can you send the shipyard definition here?

**Comment by daggs1** [comment]
> shipyard "Lionheart Custom"
> 	"Aerie (No Bays)"
> 	"Bactrian (No Bays)"
>

**Comment by warp-core** [comment]
It looks like you're not indenting the ship name lines properly.

Also, do you use Discord?

**Comment by daggs1** [comment]
> It looks like you're not indenting the ship name lines properly.
the indent is ok, the paste is bad
> Also, do you use Discord?
no

**Comment by warp-core** [comment]
Try using \`\`\` before and after the data instead of `>` so the formatting doesn't get ruined. That's how I pasted the example here.

**Comment by daggs1** [comment]
```
shipyard "Lionheart Custom"
	"Aerie (No Bays)"
	"Bactrian (No Bays)"

**Comment by warp-core** [comment]
Can you send the ship definitions the same way?

**Comment by daggs1** [comment]
```
ship "Aerie" "Aerie (No Bays)"
	attributes
		"cost" 4900000
		"cargo space" 160
		"drag" 4.1
		"outfit space" 390
		"weapon capacity" 150
		"engine capacity" 95
	-bay
	-bay

ship "Bactrian" "Bactrian (No Bays)"
	attributes
		"cost" 24640000
		"cargo space" 695
		"drag" 16.1
		"outfit space" 740
		"weapon capacity" 300
		"engine capacity" 180
	-bay
	-bay
	-bay

**Comment by warp-core** [comment]
You can't remove the bays like that (I'm not sure it's possible to remove bays in a variant, but I'll get back to you on that).

You need to give the ships `category`s like the original definitions had.
Alternatively, use `add attributes` and you won't need to set the categories here again.

**Comment by daggs1** [comment]
I added additional code that removes the bays.

I'll look into the category part
thanks

**Comment by daggs1** [comment]
I can see them now, but they are missing some parts like cost and foul, also the name doesn't differs
but I've made advancement. thanks.

**Comment by daggs1** [comment]
can I add the original's attributes and then update the values of some? like this?
```
ship "Aerie" "Aerie (No Bays)"
	add attributes
	attributes
		"cost" 4900000
		"cargo space" 160
		"outfit space" 390
	-bay
	-bay

**Comment by warp-core** [comment]
`add attributes` does not tell it to get the attributes from the original. It does that automatically.
`add attributes` tells it to add the attribute values you define below it to the attributes of the base ship.
What you are doing with just `attributes` is overriding the base attributes, which is probably where you're issues are coming from.
See the Auxiliary variants and a variant of the Korath Raider Mark XLII for examples of adding attributes in a variant.

**Comment by daggs1** [comment]
great, another inch closer, I do not however understand the cost.
I assume there the cost entry is only the hull and the rest is on the outfits.
so I assume I can set negative value to the cost and it will be reduced?
if so, I tried it, got the same cost but with (112%) along side it.
what am I missing?

**Comment by leklachu** [comment]
I've encountered a similar thing playing round with variant ships. It seems the game, for most of how it's set up, treats a variant as the base ship with different outfits... and tweaks to that ship as a minor extra. Having bought/received a ship, the variant name, and the description, are not saved in your savefile, but the practical details of the ship are.

When selling or buying, it seems to calculate the cost of the hull based off the cost which is listed in your savefile, and the depreciation in your savefile, but compare that to the 'original' - meaning the base ship definition (not the variant) - to give the %. Maybe this could be considered a bug. The name is also always the name of the base ship; the variant name is just for referencing it in the datafiles, not for the player to see.

(You're correct the cost when buying/selling is the cost of hull plus cost of outfits.)  
(And sorry, I don't know the engine code behind this; and haven't played around with variant costs since a couple of months ago.)

For reference, Marauder ships are all freshly-defined ships, separate from the vanilla, and their engines/weapons variants don't change the price from the base marauder version.

**Comment by daggs1** [comment]
I'll try to look into the cost engine.
regarding the name, it will be nice if I can mark the variant as such, for example, append to the description.
do you know if that is possible?
I will be happy to get an explanation on the % part.

**Comment by quyykk** [comment]
Negative cost is not really supported; the % is wrong because the ship loses value and for negative costs the value just becomes even more negative, hence the value over 100%.

**Reply by daggs1** [reply]
> Negative cost is not really supported; the % is wrong because the ship loses value and for negative costs the value just becomes even more negative, hence the value over 100%.

is there plan to support it?

**Reply by quyykk** [reply]
Negative cost doesn't really make sense, so I think the answer is no.

**Reply by daggs1** [reply]
I understand, I assume that maybe having internal bays as outfits is more logical. imho at least

**Comment by warp-core** [comment]
Variants are really intended just for setting different outfit loadouts without needing to redefine the whole ship.
I do have a PR open that would allow you to define a 'display name' for a variant so that the ship has a different model name displayed to the player, when it's done.

**Comment by leklachu** [comment]
The description in the shop can be changed. Look into data/human/marauders for perhaps the most instructive examples: because the Marauders are variants defined as new ships, but within those are extra engines/weapons variants. Once you buy the ship, however, the description is lost. You can never see the description for a ship you own, only find the same in a shop that sells that ship.

--- rest of comment snipped because warp-core answered it while I was writing :-D ---

**Reply by daggs1** [reply]
will look, thanks

**Comment by daggs1** [comment]
it is code I've added as part of my feature

 
 

Sent: Sunday, August 14, 2022 at 3:29 AM
